RFC: WIP: JSON endpoint for modifying the global read-only state

caio opened the pull request:
Oy,

This branch allows us to play with `EXPR_GLOBAL_STATE_RO` outside of the `search-many` context. There is plenty I'm not really happy with yet (expressed as `FIXME` comments in the files):
1. We replace the global state every single time. If we are pumping too much data, this is less than ideal. Still need to check if it's worth to use the read-write state as a temporary store
2. Clojure's hash-map to java.util.HashMap everywhere. It's rather silly. We should either completely remove it or do a proper conversion (including values).
3. Using `GET` because lazy :lollipop: 

Anyway, just getting the conversation started, I don't expect this to be merged as-is. Bash away! Also please bash my clojure-fu, I'm learning it as I write this :-)

Cheers

caio committed:
Introduce json endpoints for modifying the global state

This patch exposes 3 new GET(*) endpoints:

 1. /_state/ro_merge
 2. /_state/ro_deep_merge
 3. /_state/ro_rename

The objective is allowing modification of the state in a easier manner
than doing it via clj-eval.

*: Just because less typing. Will make it POST. Maybe.

caio committed:
Add very simple tests for bzzz.state

caio committed:
Make /_state/ro_rename_key actually work via HTTP

This patch casts from_key and to_key into keywords. This is needed
because these parameters come from a json object _value_, so they
parsed as strings.

This wasn't working:

    $ curl -XGET 'http://localhost:3000/_state/ro_merge' -d '{ "data": { "foo": 42 } }' -s

    $ curl -XGET 'http://localhost:3000/_state/ro_rename_key' -d '{ "from_key": "foo", "to_key": "bar" }' -s

Now it does :-)

Test case to maybe avoid regressions included.

jackdoe commented:
nice \o/
clojure is so much fun!
looks good, we will probably add a whole bunch of other /_state/ endpoints, i think GET is good for it
like you said in the FIXME, no need for those HashMap conversions:

```
diff --git a/src/bzzz/state.clj b/src/bzzz/state.clj
index 48d5efe..2181a6e 100644
--- a/src/bzzz/state.clj
+++ b/src/bzzz/state.clj
@@ -8,9 +8,6 @@
 ;;       If this becomes a problem, we should consider a flow where we
 ;;       write to the RW state and just merge into RO once the data pumping
 ;;       is complete.
-;; FIXME All these conversions to java.util.HashMap seem really wasteful and
-;;       make it all kinda silly, especially because it doesn't convert the
-;;       values, so it probably beats the purpose. Get rid of it? Fix it?
 ;; TODO Maybe add support for @hosts?

 (defn ro-merge
@@ -25,8 +22,7 @@
   The behavior is exactly the same as clojure.core/merge"
   [data]
   (TermPayloadClojureScoreQuery/replace_expr_global_state_ro
-    (java.util.HashMap. ^java.util.Map
-      (merge {} TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO data))))
+   (merge {} TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO data)))

 (defn ro-deep-merge
   "Deeply merge given data map into the global read-only state.
@@ -41,8 +37,7 @@
   Refer to bzzz.util/deep-merge for more details"
   [data]
   (TermPayloadClojureScoreQuery/replace_expr_global_state_ro
-    (java.util.HashMap. ^java.util.Map
-      (deep-merge (into {} TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO) data))))
+   (deep-merge TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO data)))

 (defn ro-rename-key
   "Rename a top-level key in the global read-only state.
@@ -56,7 +51,6 @@
   (ro-deep-merge { :tmp-dataset { :bar 5 }})
   (ro-rename :tmp-dataset :dataset)"
   [from_key to_key]
-  (let [ro-state (into {} TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO)]
-    (TermPayloadClojureScoreQuery/replace_expr_global_state_ro
-      (java.util.HashMap. ^java.util.Map
-        (rename-keys ro-state { (keyword from_key) (keyword to_key) })))))
+  (TermPayloadClojureScoreQuery/replace_expr_global_state_ro
+   (rename-keys TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO
+                { (keyword from_key) (keyword to_key) })))
```

looking at some usage patterns, it feels like having a temporary clojure atom that we can keep putting stuff in, and then just swapping it with replace_expr_global_state_ro will be needed as well, is really similar to what you do with the /merge and /key-rename, but without re-creating the maps with every operation
something like:
```
;; pseudo code
(def temporary-state (atom {}))

"/_state/ro_assoc_in_temporary"  (doseq [[k v] input] (swap! temporary-state assoc-in k v))
"/_state/ro_replace"  (TermPayloadClojureScoreQuery/replace_expr_global_state_ro @termporary-state)
;; using this approach will be really cheap, we just have to pay for new branches created from the assoc-in operations

```
so we can do
```
curl -XGET -d '[ [[some path],some-value], ["some-other-key-top-level", {"some-other-value": "111"}]'
http://localhost:3000/_stat/_ro_assoc_in_temporary
....
....
curl http://localhost:3000/_stat/_ro_replace
```

maybe we should also have "transformer" function, that will be able to remove the limitations from using json (like cannot have integer keys for example), before putting something in the state

in my opinion the current pull request is totally mergable :) it will take some time to get the semantics right, and it will also require some "usage" to get them right, so the sooner we have something to hack with the better


cheers
-b

caio committed:
Stop converting to java.util.HashMap

Phew.

caio committed:
Use the input directly instead of { "data": {...} }

caio committed:
Add support for temporary state and utility functions

This patch follows @jackdoe's suggestions in PR#1 and adds support for
writing to a temporary state and later on replacing/merging with the
global read-only state.

caio committed:
Expose the temporary state utils via GET

caio commented:
Great suggestions, thanks a lot! I've implemented the temporary state and also added the same "merge" flow I had added previously - mostly because some keys in the RO state might need a more frequent update then others.

I've yet to touch the `transformer` idea tho. I'll play with it tonight on a separate branch :)

jackdoe commented:
nice!
i think the current(and the previous one) version is mergable, with few small tweaks, but they can also be implemented after it is merged, so feel free to merge it, anyway here are my comments:

you dont need to copy the map into {} a new map in ro-rename-key and ro-deep-merge
```
(/replace_... (rename-keys TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO { (keyword from_key) (keyword to_key) }))
;;and 
(/replace_.. (deep-merge TermPayloadClojureScoreQuery/EXPR_GLOBAL_STATE_RO data)))
```
this is safe because in both cases because they will not try to change the input, and once deep-merge/rename-keys get the pointer they will do their magic with the current map and produce a new one, replace_expr_global_state_ro basically is swapping a pointer, which is atomic

the only issue i see there is that it will be nice if we wrap all those calls in a giant lock, so if we decide to update the state from multiple threads, we dont lose updates (which at the moment will happen)

```
[[:a 1], [:c 1]] in thread A
[[:b 1]] in thread B

thread A     thread B
{}                {}
{:a 1}          {:b 1}
{:c 1}
{:c 1 :b 1}
```

so `(locking temporary-state ....)` will be enough.

will also be nice to have /_stat data for time took to do the operations (https://github.com/jackdoe/bzzz/blob/master/src/bzzz/discover.clj#L75) for example `(index-stat/update-took-count index-stat/total "state-merge" (time-took t0))`

-b

caio committed:
Don't copy into {} before rename-keys or deep-merge

I put those `into {}` to work around a failure back when we used to
convert the state into a java.map.HashMap before setting it as the
new global state.

This can still be reproduced by just calling rename-keys in the original
state (a java.util.HashMap instance). We get this:

    java.lang.ClassCastException: java.util.HashMap cannot be cast to clojure.lang.IPersistentMap

This patch adds a test to the scenario and changes the original instance
of `EXPR_GLOBAL_STATE_RO` to be a `clojure.lang.PersistentArrayMap`.

caio commented:
Cool, I'm merging then! I've made it not copy into an empty map on 16b79262 (it was to workaround an issue, I "fixed" it differently now). Without this patch a `rename-keys` in the original RO state throws an exception.

To-do:

 * Endpoint to transform the data via `clj-eval`
 * Put a giant lock for `temporary-state`
